Battle refactor: No longer reset negated immunity in pokemon.update()

ascriptmaster opened the pull request:
Replace the pokemon.negateImmunity object with a new `NegateImmunity`
event that handles all the immunity negation.

Also deprecate 'IgnoreEffectiveness' in favor of having relevant moves
and items use `Effectiveness` event handlers instead.

ascriptmaster commented:
@Marty-D

Marty-D commented:
Passes the existing effectiveness tests; looks good to me! :+1: for less use of `ModifyPokemon`.

ascriptmaster commented:
Let me write up Wonder Guard tests first, just in case.

ascriptmaster commented:
Suddenly Wonder Guard.

Zarel commented on battle-engine.js:
```diff
@@ -1295,8 +1293,8 @@ BattlePokemon = (function () {
-		if (this.negateImmunity[type]) return true;
-		if (!(this.negateImmunity['Type'] && type in this.battle.data.TypeChart)) {
+		if (!this.battle.runEvent('NegateImmunity', this, type)) return true;
+		if (this.battle.runEvent('NegateImmunity', this, 'Type') || !(type in this.battle.data.TypeChart)) {
```
Two event-runs? Is there any way to get this down to one?

ascriptmaster replied:
Hmmm. Since the old `negateImmunity['Type']` was designed to only stop natural immunity and not ability-based, the other solution that came to mind was to have the relevant event handlers run the `Immunity` event inside.

But I guess it's better to have 2 events sometimes than 2 events always.

ascriptmaster commented:
Today I learned what `git commit --fixup` is and it's definitely better than having to move the lines myself

Zarel commented:
Also set `rebase.autosquash` in `git config` to save yourself some typing.

ascriptmaster commented:
It was already on. That's probably why it saved some typing in the first place. If it was off I'd probably have to move the lines anyways

ascriptmaster commented:
Anyways how does this look

Slayer95 commented:
Tests for Ring Target?

ascriptmaster commented:
Included in #1942. If those aren't sufficient then let me know.

Slayer95 commented:
Oh, they are.
Uhm, this looks too convoluted / hacky.

ascriptmaster commented:
What part of it is too convoluted/hacky... or the entire system, by adding a new Event?

ascriptmaster commented:
I might have to ask Marty some questions about Thousand Arrows' details before this gets merged.

Slayer95 commented:
Went over previous discussions with Marty and added these currently failing tests: https://github.com/Zarel/Pokemon-Showdown/commit/21aebd070e80ebefca588eac5caffdf447a89d78

If you can tweak this implementation to make them pass, I am all for it.

ascriptmaster commented:
Rebased.

I also changed the mechanics on Thousand Arrows and Iron Ball based on my understanding of what @Marty-D described. My assumptions were basically that you can't Smack Down a Grounded Pokemon, regardless of whether it's Flying-type, Levitate, etc. and Iron Ball doesn't do the neutral to Ground thing if they're grounded by another effect.

Slayer95 commented on test/simulator/moves/thousandarrows.js:
```diff
@@ -29,31 +29,31 @@ describe('Thousand Arrows', function () {
 		battle.join('p1', 'Guest 1', 1, [{species: "Zygarde", level: 10, ability: 'aurabreak', item: 'laggingtail', moves: ['thousandarrows']}]);
 		battle.join('p2', 'Guest 2', 1, [{species: "Ho-Oh", ability: 'wonderguard', item: 'ringtarget', moves: ['recover']}]);
 		battle.commitDecisions();
-		assert.notStrictEqual(battle.p2.active.hp, battle.p2.active.maxhp);
```
My bad ;-;

Slayer95 commented on test/simulator/moves/thousandarrows.js:
```diff
@@ -29,31 +29,31 @@ describe('Thousand Arrows', function () {
+		battle.choose('p1', 'move 2');
 		battle.choose('p2', 'move 2');
 		assert.strictEqual(hp, battle.p2.active[0].hp);
 	});
```
How is this test wrong?

ascriptmaster replied:
We're checking to make sure Thousand Arrows didn't actually ground the Pokemon that was holding Iron Ball previously. How are we supposed to check that it's regained its Ground immunity if we pick the move that's specifically designed to hit Pokemon that are immune to Ground?

Slayer95 commented on test/simulator/moves/thousandarrows.js:
```diff
@@ -29,31 +29,31 @@ describe('Thousand Arrows', function () {
-		battle.join('p1', 'Guest 1', 1, [{species: "Zygarde", level: 10, ability: 'aurabreak', item: 'laggingtail', moves: ['thousandarrows', 'mudslap']}]);
-		battle.join('p2', 'Guest 2', 1, [{species: "Ho-Oh", ability: 'wonderguard', item: 'ironball', moves: ['recover', 'gravity']}]);
+		battle.join('p1', 'Guest 1', 1, [{species: "Zygarde", level: 10, ability: 'aurabreak', item: 'laggingtail', moves: ['thousandarrows', 'sleeptalk']}]);
+		battle.join('p2', 'Guest 2', 1, [{species: "Ho-Oh", ability: 'shellarmor', moves: ['recover', 'gravity']}]);
```
Iron Ball + Gravity is an exact test case described by @Marty-D 
Maybe I should have chosen a better title for the test...

ascriptmaster replied:
Except, this is a Thousand Arrows test with Gravity. Iron Ball Gravity is covered by the Iron Ball test suite.
